File: backend/src/algorithms/grid.ts
import type { Grid } from "../types/puzzle.js";

const BLOCKED_CELL = "#";
const EMPTY_CELL = "";

export function normalizeGrid(grid: Grid): Grid {
  if (grid.length === 0) {
    throw new Error("The crossword grid must contain at least one row.");
  }

  const columnCount = grid[0]?.length ?? 0;

  if (columnCount === 0) {
    throw new Error("The crossword grid must contain at least one column.");
  }

  return grid.map((row, rowIndex) => {
    if (row.length !== columnCount) {
      throw new Error(`Grid row ${rowIndex + 1} has an inconsistent length.`);
    }

    return row.map((cell, columnIndex) => {
      const normalizedCell = cell.trim().toUpperCase();

      if (
        normalizedCell !== BLOCKED_CELL &&
        normalizedCell !== EMPTY_CELL &&
        !/^[A-Z]$/.test(normalizedCell)
      ) {
        throw new Error(
          `Invalid cell at row ${rowIndex + 1}, column ${columnIndex + 1}.`,
        );
      }

      return normalizedCell;
    });
  });
}

export function normalizeWords(words: string[]): string[] {
  const normalizedWords = words
    .map((word) => word.trim().toUpperCase())
    .filter(Boolean);

  if (normalizedWords.some((word) => !/^[A-Z]+$/.test(word))) {
    throw new Error("Candidate words may contain letters only.");
  }

  return [...new Set(normalizedWords)];
}

export function isOpenCell(grid: Grid, row: number, column: number): boolean {
  return grid[row]?.[column] !== undefined && grid[row]?.[column] !== BLOCKED_CELL;
}

File: backend/app/algorithms/grid.py
import re
from typing import List

def is_open_cell(grid: List[List[str]], row: int, col: int) -> bool:
    if row < 0 or row >= len(grid):
        return False
    if col < 0 or col >= len(grid[row]):
        return False
    cell = grid[row][col]
    return cell != "#"

def normalize_grid(grid: List[List[str]]) -> List[List[str]]:
    normalized = []
    for r_idx, row in enumerate(grid):
        new_row = []
        for c_idx, cell in enumerate(row):
            if not isinstance(cell, str):
                raise ValueError(f"Invalid cell at row {r_idx + 1}, column {c_idx + 1}.")
            val = cell.strip().upper()
            if val == "#":
                new_row.append("#")
            elif val == "":
                new_row.append("")
            elif re.match(r"^[A-Z]$", val):
                new_row.append(val)
            else:
                raise ValueError(f"Invalid cell value '{cell}' at row {r_idx + 1}, column {c_idx + 1}.")
        normalized.append(new_row)
    return normalized

def normalize_words(words: List[str]) -> List[str]:
    normalized = []
    for word in words:
        w = re.sub(r"[^A-Z]", "", word.strip().upper())
        if w:
            normalized.append(w)
    return normalized
